### PR TITLE
Force manually service restart on upgrade

### PR DESCRIPTION
**Pull request by cstoidner**
## Proposed changes

If a user wants to update an existing thin-edge instance on it's device, he must actually stop services manually before any upgrade. It's better to use `tedge disconnect` instead of stopping services directly with `systemctl`, as the knowledge which services to handle is implemented in these commands. Finally it is consequent to let the user afterwards also start services also manually again (using `tedge connect`) instead of trying to start in the DEB maintainer scripts.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/> None

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by didier-wenzek** [comment]
> I'm a bit lost regarding the plan beyond that PR.

Okay, I have a better understanding after your latest [comment on the self-update plugin](https://github.com/thin-edge/thin-edge.io/pull/872#discussion_r809265929).

So the plan is:

* The cloud admin user trigger an update of thin-edge.
* The agent receives this request via the mapper and forwards it the self-update plugin.
* The self-update plugin
    * downloads the packages.
    * calls `tedge disconnect`, stopping the agent and the mappers
    * install the packages
    * let somewhere a message for the agent to tell the status of the self-update
    * calls `tedge connect`, restarting the new agent and mappers using the new tedge.
    * makes sure the connection is back with `tedge connect <cloud> test`
    * and dies.
 * On restart, the agent read the message from the plugin and forwards this status to the cloud via the mapper.

Is this correct?

What I understand now sounds as a good direction of work. Notably, I think we can then address corner cases as "what if `tedge connect c8y --test` fails". The question is still open but we can at least detect the issue while this would have been trickier in the post-install scripts.

File: docs/src/howto-guides/002_installation.md [hunk 1]
```diff
@@ -39,11 +39,12 @@ When all dependencies are in place you can proceed with installation of `thin-ed
 
 If you already have `thin-edge.io` on your device to upgrade `thin-edge.io` follow the steps below, there is no need to remove old version.
 
-> Note: To successfully upgrade `thin-edge.io` all thin-edge.io components should be stopped, eg: `tedge-mapper` and `tedge-agent`:
+> Note: To successfully upgrade `thin-edge.io` all thin-edge.io components must stopped before upgrade, and started again afterwards. Easiest was is to use commands `tedge disconnect <cloud>` and `tedge connect <cloud>`. Example:
```
Comment (didier-wenzek) [inline docs/src/howto-guides/002_installation.md:42]:
```suggestion
> Note: To successfully upgrade `thin-edge.io` all thin-edge.io components must be stopped before upgrade, and started again afterwards. The simpler is to use the commands `tedge disconnect <cloud>` and `tedge connect <cloud>`. Example:
```

**Review by albinsuresh — COMMENTED** [review]
Since the expectation from the user is to have all the services shutdown before an upgrade, wouldn't it be better if we validate that too from our debian maintainer scripts? May be, from the `prerm` or `preinst` commands which will fail if they detect that these services are running during an update?

File: docs/src/howto-guides/002_installation.md [hunk 2]
```diff
@@ -39,11 +39,12 @@ When all dependencies are in place you can proceed with installation of `thin-ed
 
 If you already have `thin-edge.io` on your device to upgrade `thin-edge.io` follow the steps below, there is no need to remove old version.
 
-> Note: To successfully upgrade `thin-edge.io` all thin-edge.io components should be stopped, eg: `tedge-mapper` and `tedge-agent`:
+> Note: To successfully upgrade `thin-edge.io` all thin-edge.io components must be stopped before upgrade, and started again afterwards. The simpler is to use the commands `tedge disconnect <cloud>` and `tedge connect <cloud>`. Example:
```
Comment (rina23q) [inline docs/src/howto-guides/002_installation.md:42]: I really support this change. As of now, a user needs to stop maximum 4 services to upgrade all packages. `tedge disconnect` does all by just one command. It's more user-friendly. It is meaningful that we give this option to users in our user-guide.

File: configuration/debian/tedge_agent/postinst [hunk 3]
```diff
@@ -28,20 +28,6 @@ install -g tedge-agent -o tedge-agent -m 754 -d /etc/tedge/.agent
...
-        # start and enable tedge-agent
-        systemctl start tedge-agent.service
-        systemctl enable tedge-agent.service
-
-        # start and enable tedge-mapper-sm-c8y
-        systemctl start tedge-mapper-sm-c8y.service
-        systemctl enable tedge-mapper-sm-c8y.service
-    fi
-fi
-
```
[the remark this answers is not in the document]
Comment (rina23q) [reply configuration/debian/tedge_agent/postinst:44]: For me, it depends on what we recommend users when they upgrade packages manually.

As of now, we write "user should stop the services individually (not using disconnect)", so the story "bridge config exists (user already connected c8y)", AND "user wants to upgrade packages" is valid.

However, if we change the user-guide to recommend users "user should run `tedge disconnect c8y` before upgrading packages", that story is invalid.

**Review by rina23q — APPROVED** [review]
I approve it based on my preference.

File: configuration/debian/tedge_agent/postinst (hunk 3 above)
[the remark this answers is not in the document]
Comment (makr11st) [reply configuration/debian/tedge_agent/postinst:44]: This change has not much to do with upgrade operation.
Removing these lines would not enable and start the services on fresh installations...

[the remark this answers is not in the document]
Comment (rina23q) [reply configuration/debian/tedge_agent/postinst:44]: For fresh installation, I can't find any reason to start tedge_mapper and tedge_agent right after the installation. I believe `tedge connect` is responsible to it.

**Review by rina23q — APPROVED** [review]

**Comment by cstoidner** [comment]
No longer applicable, since restart behaviour on upgrade significantly changed since 0.8.0, see https://github.com/thin-edge/thin-edge.io/issues/1466
